File: grind75-02-valid-parentheses.py
# https://leetcode.com/problems/valid-parentheses/
class Solution:
    # Time complexity: O(n)
    # Space complexity: O(n)
    def isValid(self, s: str) -> bool:
        PAREN_MAP = { '(': ')', '[': ']', '{': '}' }
        open_paren_stack = []
        for char in s:
            if char in PAREN_MAP:
                open_paren_stack.append(char)
            elif not open_paren_stack or char != PAREN_MAP[open_paren_stack.pop()]:
                return False
        return not open_paren_stack

File: grind75-02-valid-parentheses.ts
// https://leetcode.com/problems/valid-parentheses/
// 80th percentile
function isValid(s: string): boolean {
    const BRACKETS = {
        '(': ')',
        '[': ']',
        '{': '}',
    }
    let closeBracketStack: string[] = []
    for (let bracket of s) {
        let closeBracket = BRACKETS[bracket]
        if (closeBracket !== undefined) {
            closeBracketStack.push(closeBracket)
        } else if (bracket !== closeBracketStack.pop()) {  // also handles undefined pop
            return false
        }
    }
    return !closeBracketStack.length
}